### DISCUSSION TITLE
Custom validators for enums (V2)

### DISCUSSION THREAD

**Discussion by Galarzaa90** [pydantic/pydantic#6466] [Question] [answered]
Just started migrating to Pydantic V2, but something that I'm struggling with is with my enum classes.

In short, I'm trying to achieve two things:
- Deserialize from member's name.
  I was achieving this before using `__get_validators__`.
- Serialize as member's name
  I was doing this using `json_encoders`
 
  ```py
      class Config:
          json_encoders = {NumericEnum: lambda g: g.name}
  ```

**Comment by adriangb** [comment]
Would something like this work for your use case? https://docs.pydantic.dev/latest/usage/types/custom/#adding-validation-and-serialization

**Reply by Galarzaa90** [reply]
So I tried using `Annotated` and that worked well for serializing it. It is a hassle that I need to update all models that use enums. But it is a good alternative if I don't find anything else.

I also tried using the @model_serializer:

```py
@model_serializer
    def ser_model(self) -> str:
        return self.name
```

But didn't work, most likely because enum classes are not inheriting `BaseModel`.

What did work to serialize is adding this:

```py
    @classmethod
    def __get_pydantic_core_schema__(
            cls, source: Type[Any], handler: GetCoreSchemaHandler
    ) -> core_schema.CoreSchema:
        return core_schema.general_after_validator_function(
            cls.validate,
            core_schema.int_schema(),
            serialization=core_schema.to_string_ser_schema(),
        )

    @classmethod
    def validate(cls, v: int, _info):
        return try_enum(cls, v).name
```

I'm still trying to figure out deserialization, but didn't have much chances to test today.

**Reply by samuelcolvin** [reply]
You can also customize serialization using Annotated, use https://docs.pydantic.dev/latest/api/functional_serializers/#pydantic.functional_serializers.PlainSerializer

I'm on my phone at the moment so can't give an example.

**Reply by adriangb** [reply]
How about something like this:

```python
from enum import Enum, auto
from typing import Any

from pydantic_core import CoreSchema, core_schema

from pydantic import BaseModel, GetCoreSchemaHandler, GetJsonSchemaHandler
from pydantic.json_schema import JsonSchemaValue


class MyEnumBase(Enum):
    @classmethod
    def __get_pydantic_core_schema__(cls, _source_type: Any, _handler: GetCoreSchemaHandler) -> CoreSchema:
        return core_schema.no_info_after_validator_function(
            lambda x: getattr(cls, x),
            core_schema.str_schema(),
            serialization=core_schema.plain_serializer_function_ser_schema(lambda x: x.name),
        )

    @classmethod
    def __get_pydantic_json_schema__(cls, _core_schema: CoreSchema, _handler: GetJsonSchemaHandler) -> JsonSchemaValue:
        return {'enum': [m.name for m in cls], 'type': 'string'}


class MyEnum(MyEnumBase):
    a = auto()
    b = auto()


class Model(BaseModel):
    x: MyEnum


print(Model.model_validate({'x': 'a'}))
print(Model.model_validate({'x': 'a'}).model_dump())
print(Model.model_json_schema())
```

The same thing could be achieved with `Annotated` to but it sounds like you wanted more of a base class approach.

**Reply by Galarzaa90** [reply, marked as answer]
@adriangb that worked! Well, almost ,just had to change `core_schema.str_schema()` to `core_schema.any_schema()`.

`str_schema` would make deserialization work, but not serialization. (couldn't even try to create a model directly)
`int_schema` would make serialization work, but not deserialization.

**Reply by Jan-Jasek** [reply]
@adriangb 
Just wondering, what would be an approach to achieve the same with `Annotated` ? It works fine for single Enum, but let's say I have `n` enumeration classes defined and I want to define single serialization for all of them.

**Reply by adriangb** [reply]
Yep!

```python
from enum import Enum, auto
from typing import Annotated, Type, TypeVar

from pydantic_core import CoreSchema, core_schema

from pydantic import BaseModel, GetCoreSchemaHandler, GetJsonSchemaHandler
from pydantic.json_schema import JsonSchemaValue


class EnumSerializationCustomizer:
    def __get_pydantic_core_schema__(self, enum_cls: Type[Enum], _handler: GetCoreSchemaHandler) -> CoreSchema:
        def customize_json_schema(_core_schema: CoreSchema, _handler: GetJsonSchemaHandler) -> JsonSchemaValue:
            return {'enum': [m.name for m in enum_cls], 'type': 'string'}

        s = core_schema.no_info_after_validator_function(
            lambda x: getattr(enum_cls, x),
            core_schema.str_schema(),
            serialization=core_schema.plain_serializer_function_ser_schema(lambda x: x.name),
        )
        s.setdefault('metadata', {}).setdefault('pydantic_js_functions', []).append(customize_json_schema)
        return s


class MyEnum(Enum):
    a = auto()
    b = auto()


EnumType = TypeVar('EnumType', bound=Enum)

SerByName = Annotated[EnumType, EnumSerializationCustomizer()]


class Model(BaseModel):
    x: SerByName[MyEnum]


print(Model.model_validate({'x': 'a'}))
print(Model.model_validate({'x': 'a'}).model_dump())
print(Model.model_json_schema())
```

But this requires using the `'metadata'` thing which is kind of a hack (and we don't promise long-term support).
You could also use a mixin so you don't have code duplication, but you'd still need to modify / subclass each one of the enum classes.

**Reply by dkbarn** [reply]
Feature request: Could we have the target type of the value being validated accessible from one of the arguments passed to a `WrapValidator` function? Right now that function receives a `handler` argument and optionally an `info` argument, but neither of these objects contain the value's type. This would make the sort of validation required here actually possible with a simple `WrapValidator`.

Alternatively, could there simply be a new option to `ConfigDict` called `enums_by_name`? This is how SQLAlchemy, for example, deals with supporting both enum-by-name and enum-by-value while making it very easy for the developer.

**Comment by mzamini92** [comment]
Does this example work for you?
```
from enum import Enum
from pydantic import BaseModel, BaseConfig

class NumericEnum(Enum):
    ONE = 1
    TWO = 2
    THREE = 3

class MyModel(BaseModel):
    number: NumericEnum

class Config(BaseConfig):
    json_encoders = {NumericEnum: lambda g: g.name}

my_model = MyModel(number=NumericEnum.ONE)

print(my_model.json())
```
outputs:
```
{"number": 1}
```

**Reply by Galarzaa90** [reply]
There is no `json_encoders` configuration in Pydantic V2

**Comment by brthor** [comment]
A note for anyone else landing here from google.

The second part: "Serialize as member's name", can be achieved neatly with `PlainSerializer`.

```python
import pydantic, enum
from typing import Annotated

EnumNameSerializer = pydantic.PlainSerializer(
    lambda e: e.name,
    return_type='str',
    when_used='always')

class MyEnum(enum.Enum):
    name1 = 0
    name2 = 1

class MyModel(pydantic.BaseModel):
    the_enum: Annotated[MyEnum, EnumNameSerializer]

print(MyModel(the_enum=MyEnum.name2).model_dump_json())
```

This outputs

```
{"the_enum":"name2"}
```

**Comment by dkbarn** [comment]
From the title of this discussion, I had high hopes that a solution would be provided for **validating** enums by name, but sadly the conversation seems to be focused on **serializing** enums by name.

Has anyone worked out how to validate / deserialize an enum by name in Pydantic V2?

**Reply by NiallRees** [reply]
Does this do what you need?

```
    @classmethod
    def __get_pydantic_core_schema__(
        cls, _source_type: Any, _handler: GetCoreSchemaHandler
    ) -> CoreSchema:
        return core_schema.no_info_after_validator_function(
            cls.validate,
            core_schema.any_schema(),
            serialization=core_schema.plain_serializer_function_ser_schema(
                lambda x: x.name
            ),
        )

    @classmethod
    def validate(cls, v: Any):
        if isinstance(v, cls):
            return v
        elif isinstance(v, str):
            try:
                return cls[v]
            except KeyError:
                raise ValueError(f"Invalid {cls.__name__}: {v}")
        else:
            raise ValueError(f"Unexpected type: {type(v)}")

```
